### PR TITLE
tool/go: add wrapper to download and use go.toolchain.rev go version.

### PR DESCRIPTION
**Pull request by maisem**
Also update build_dist.sh and build_docker.sh to use the wrapper.

Updates tailscale/corp#3385

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

**Comment by crawshaw** [comment]
Isn't this reinventing gocross?

**Comment by maisem** [comment]
Doesn't gocross do a bunch more stuff?

**Review by crawshaw — APPROVED** [review]
LGTM in the spirit of forward progress, but hopefully we can stop duplicating shell scripts and use one piece of code to do this one day.

**Comment by bradfitz** [comment]
I have a few comments when back at a computer in hour or two

File: build_dist.sh [hunk 1]
```diff
@@ -45,4 +45,4 @@ EOF
 	exit 0
```
Comment (bradfitz) [inline build_dist.sh:45]: typo in commit message title: `go.toolcahin.rev`

File: build_docker.sh [hunk 2]
```diff
@@ -19,6 +19,9 @@
 
 set -eu
 
+# Use tailscale-go.
```
Comment (bradfitz) [inline build_docker.sh:22]:
```
# Use the "go" binary from the "tool" directory (which is github.com/tailscale/go)
```

File: tool/go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                BUILD="$GOOS"
+            else
+                BUILD="$GOOS-$ARCH"
+            fi
+            curl -f -L -o "$archive.new" "https://github.com/tailscale/go/releases/download/build-${REV}/${BUILD}.tar.gz"
+            rm -f "$archive"
+            mv "$archive.new" "$archive"
+    fi
+
+    log "Extracting go '$REV'" >&2
```
Comment (bradfitz) [inline tool/go:69]:
```
Extracting tailscale/go rev '$REV'
```

File: tool/go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        # already ok
+        log "Go toolchain '$REV' already extracted."
+        return 0
+    fi
+
+    rm -f "$archive.new" "$TOOLCHAIN.extracted"
+    if [ ! -e "$archive" ]; then
+            log "Need to download go '$REV'."
+            if [ "$ARCH" = "amd64" ]; then
+                # For historic reasons, amd64 binaries don't have the arch in the name.
```
Comment (bradfitz) [inline tool/go:59]:
```
# For historic reasons, the tailscale/go amd64 release artifacts don't
# have the arch in their name.
```

File: tool/go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    log "Extracting go '$REV'" >&2
+    log "  into '$TOOLCHAIN'." >&2
+    rm -rf "$TOOLCHAIN"
+    mkdir -p "$TOOLCHAIN"
+    (cd "$TOOLCHAIN" && tar --strip-components=1 -xf "$archive")
+    echo "$REV" >$mark
+}
+
+if [ "${REV}" = "SKIP" ]; then
+    # Use whichever go is available
```
Comment (bradfitz) [inline tool/go:78]: We should also do this on non-Linux/non-Darwin or non-amd64/non-arm64.

e.g. linux/386 or linux/arm?

File: tool/go [hunk 6]
```diff
@@ -0,0 +1,85 @@
+#!/bin/sh
+#
+# This script acts like the "go" command, but uses Tailscale's
+# currently-desired version, downloading it first if necessary.
```
Comment (bradfitz) [inline tool/go:4]: expand this comment to link to https://github.com/tailscale/go too

**Review by bradfitz — APPROVED** [review]

**Pull request merged**